File: 2018-kakao-blind-recruitment/Q2.1차-다트게임.py
import re


def solution(dartResult):
    bonus_dic = {'S': '1', 'D': '2', 'T': '3'}
    option_dic = {'*': '2', '#': '-1'}

    score_list = []

    regex = re.compile('\d{1,2}[SDT][*#]?')
    dart_result_list = re.findall(regex, dartResult)

    for dart_result in dart_result_list:
        score_search = re.search('\d{1,2}', dart_result)[0]
        bonus_search = re.search('[SDT]', dart_result)[0]
        option_search = re.search('[*#]', dart_result)

        result_score = int(score_search) ** int(bonus_dic[bonus_search])

        if option_search:
            option = option_search[0]
            option_num = int(option_dic[option])

            if option == '*' and score_list:
                score_list[-1] *= option_num
            result_score *= option_num

        score_list.append(result_score)

    return sum(score_list)

File: 2018-kakao-blind-recruitment/Q2.1차-다트게임.js
function matchOf(str, regex) {
	const res = str.match(regex);
	return res ? res[0] : null;
}

function solution(dartResult) {
	const scoreList = [];
	const bonusObj = { S: 1, D: 2, T: 3 };
	const optionObj = { '*': 2, '#': -1 };

	const handleOption = (score, option) => {
		if (option === '*' && scoreList.length) {
			scoreList[scoreList.length - 1] *= optionObj[option];
		}

		return score * optionObj[option];
	};

	const dartResultRegex = /\d{1,2}[SDT][*#]?/g;
	const scoreRegex = /\d{1,2}/;
	const bonusRegex = /[SDT]/;
	const optionRegex = /[*#]/;

	const resultList = dartResult.match(dartResultRegex);

	for (const result of resultList) {
		let score = +matchOf(result, scoreRegex);
		const bonus = matchOf(result, bonusRegex);
		const option = matchOf(result, optionRegex);

		score **= bonusObj[bonus];

		if (option) {
			score = handleOption(score, option);
		}
		scoreList.push(score);
	}

	return scoreList.reduce((a, b) => a + b);
}
